google/cloud/spanner/internal/time.cc

```diff
@@ -29,156 +28,161 @@ namespace spanner {
+  return std::chrono::seconds(proto.seconds()) +
+         std::chrono::nanoseconds(proto.nanos());
 }
 
 //
 // Timestamp
 //
 
-google::protobuf::Timestamp ToProto(time_point tp) {
-  std::time_t t = system_clock::to_time_t(tp);
-  time_point ttp = system_clock::from_time_t(t);
-  auto ss = std::chrono::duration_cast<nanoseconds>(tp - ttp);
-  if (ss.count() < 0) {
-    t -= 1;
-    ss += seconds(1);
+namespace {
+
+// RFC3339 "date-time" prefix (no "time-secfrac" or "time-offset").
+constexpr auto kTimeFormat = "%Y-%m-%dT%H:%M:%S";
+
+// Split a Timestamp into a seconds-since-epoch and a (>=0) subsecond.
+std::pair<std::chrono::seconds, Timestamp::duration> SplitTime(Timestamp ts) {
+  auto e = std::chrono::system_clock::from_time_t(0);
+  auto d = ts - std::chrono::time_point_cast<Timestamp::duration>(e);
+  auto s = std::chrono::duration_cast<std::chrono::seconds>(d);
+  auto ss = d - std::chrono::duration_cast<Timestamp::duration>(s);
+  if (ss < Timestamp::duration(0)) {
+    s -= std::chrono::seconds(1);
+    ss += std::chrono::seconds(1);
   }
-  google::protobuf::Timestamp proto;
-  proto.set_seconds(t);
-  proto.set_nanos(ss.count());
-  return proto;
+  return {s, ss};
 }
 
-time_point FromProto(google::protobuf::Timestamp const& proto) {
-  nanoseconds ss(proto.nanos());
-  auto sub = std::chrono::duration_cast<system_clock::duration>(ss);
-  return system_clock::from_time_t(proto.seconds()) + sub;
+// Combine a seconds-since-epoch and a subsecond into a Timestamp.
+Timestamp CombineTime(std::chrono::seconds s, Timestamp::duration ss) {
+  auto e = std::chrono::system_clock::from_time_t(0);
+  return std::chrono::time_point_cast<Timestamp::duration>(e) + (s + ss);
 }
 
-namespace {
-
-// A duration capable of holding subsecond values at high precision.
-using femtoseconds = std::chrono::duration<std::int64_t, std::femto>;
-
-// Convert a std::time_t into a Zulu std::tm.
+// Convert a seconds-since-epoch into a Zulu std::tm.
 //
 // See http://howardhinnant.github.io/date_algorithms.html for an explanation
 // of the calendrical arithmetic in ZTime() and TimeZ().  For quick reference,
 // March 1st is used as the first day of the year (so that any leap day occurs
 // at year's end), there are 719468 days between 0000-03-01 and 1970-01-01,
 // and there are 146097 days in the 400-year Gregorian cycle (an era).
-std::tm ZTime(std::time_t const t) {
-  std::time_t sec = t % (24 * 60 * 60);
-  std::time_t day = t / (24 * 60 * 60);
-  if (sec < 0) {
-    sec += 24 * 60 * 60;
-    day -= 1;
+//
+// All the civil-time code assumes the proleptic Gregorian calendar, with
+// 24-hour days divided into 60-minute hours and 60-second minutes.
+std::tm ZTime(std::chrono::seconds s) {
+  using rep = std::chrono::seconds::rep;
+  constexpr auto kSecsPerDay = std::chrono::hours::period::num * 24;
+  using days = std::chrono::duration<rep, std::ratio<kSecsPerDay>>;  // C++20
```

Comment:
nit: what does the `// C++20` comment mean?  That you miss that feature from C++20?

Reply:
It was supposed to hint that the std::chrono::days duration exists in C++20, but the short hint is indeed more confusing that it is worth, and nor is it useful to go to any lengths to describe the future.    So, deleted.